service/integration/open_pbs/README.rst

```diff
@@ -28,57 +36,250 @@ correct the power limit setting if it was not restored as part of the previous
 job. More specifically, if a new power limit is not requested in the next job
 involving the affected nodes (and the save file is present), the hook will
 take the opportunity to restore the power limit setting during the prologue so
-the new job is not affected by a prior power limit setting.
+the new job is not affected by a prior power limit setting. The save file is
+removed when the power limits are restored (regardless of whether the prologue
+or epilogue handles the restore operation).
+
+Automatic Power Limits
+----------------------
+The goal of the automatic power limiter is to provide additional scheduling
+flexibility when the cluster is power constrained. If the cluster power cap is
+sufficiently low, then jobs may not begin executing even when sufficient
+non-power resources are available on idle nodes. At a high level,
+this automatic limiter attempts to set the lowest acceptable power cap to
+each job (based on modeled power-performance relationships and a desired
+slowdown limit) until the scheduler offers to allocate resources to a job. When
+the scheduler offers to start the job, the automatic limiter opportunistically
+increases power to that job if there is additional unallocated power available
+in the cluster.
+
+Automatic power limits are selected through handlers for ``queuejob`` and
+``runjob`` PBS events. The ``queuejob`` event handler initially sets the
+``geopm-job-power-limit`` of each submitted job to be the lowest tolerable
+power limit for that job, according to the configured maximum job slowdown and
+the job's selected power-performance model. Once the PBS
+scheduler has determined that there are sufficient resources (power and
+otherwise) to execute the job, the ``runjob`` event handler is triggered. That
+event handler opportunistically increases the power limit for that job, up to
```

Comment:
Would the runjob event handler ever reduce the power limit back down to the original lower bound to make space for a new job?

Reply:
This is an escaped part of documentation from my latest updates, so I'll remove that note. The rest of this comment is just additional context.

I originally implemented the capability described here but I removed it after realizing that it does not play nice with backfilling. As far as I can tell, it doesn't cause any constraint violations, but it would potentially cause a candidate fill job to lose its backfill spot in the schedule, which breaks a key motivation to get users to opt in to do this.

Example scenario: There is a power-and-time hole in the schedule, which can be assigned to a particular fill job that was assigned a low power cap on queue entry. Now that it is identified as a viable fill job, we execute the runjob event handler, which increases the power assigned to this job to use whatever slack power **currently** exists. But even if there is currently sufficient power, that may not be true for all times between now and the job's time limit. The scheduler will re-evaluate viability of this job after changing the job's resource requirements. If the increased power cap causes an expected delay for the top job, then this job will no longer be used as a fill job.

I think we could avoid that problem if it was possible to ask the scheduler whether the resource change would make us lose our spot, but I haven't figured out a way to do that. So this implementation just keeps the initial cap set by queuejob.